deepchem/models/jax_models/jax_model.py

```diff
@@ -0,0 +1,365 @@
+import numpy as np
+import time
+import logging
+try:
+  from collections.abc import Sequence as SequenceCollection
+except:
+  from collections import Sequence as SequenceCollection
+
+from deepchem.data import Dataset
+from deepchem.models.models import Model
+from deepchem.models.losses import Loss
+from deepchem.models.optimizers import Optimizer
+from typing import Any, Callable, Iterable, List, Optional, Tuple, Union
+from deepchem.utils.typing import LossFn
+
+# JAX depend
+import jax.numpy as jnp
+import jax
+import haiku as hk
+import optax
+
+logger = logging.getLogger(__name__)
+
+
+class JaxModel(Model):
+  """This is a DeepChem model implemented by a Jax Model
+
+  Here is a simple example of that uses JaxModel to train a
+  Haiku (JAX Nueral Network Library) based model on deepchem
+  dataset.
+
+  >> def f(x):
+  >>   net = hk.nets.MLP([512, 256, 128, 1])
+  >>   return net(x)
+  >> model = hk.without_apply_rng(hk.transform(f))
+  >> rng = jax.random.PRNGKey(500)
+  >> x, _, _, _ = next(iter(train_dataset.iterbatches(batch_size=256)))
+  >> params = model.init(rng, x)
+  >> j_m = JaxModel(model, params, 256, 0.001, 100)
+  >> j_m.fit(train_dataset)
+
+  All optimizations will be done using the optax library.
+  """
+
+  def __init__(self,
+               model,
```

Comment:
Type signature for `model` needed

Reply:
I still haven't decided on the model signature, because Jax in its true forms uses pure functions to calculate gradients. Hence mostly likely I will call it `Callable` type. 

There are many popular Jax Neural Network Libraries like [`Haiku`](https://github.com/deepmind/dm-haiku) and [`Flax`](https://github.com/google/flax), and im trying to build it library agnostic. In my current experiments, I have been using `Haiku` models.

Reply:
`Callable` sounds good here. Could also do `Union[...,...]` with each framework that is supported.